java/ql/lib/semmle/code/java/regex/RegexTreeView.qll

```diff
@@ -0,0 +1,1031 @@
+}
+
+/**
+ * A regular expression term, that is, a syntactic part of a regular expression.
+ */
+class RegExpTerm extends RegExpParent {
+  Regex re;
+  int start;
+  int end;
+
+  RegExpTerm() {
+    this = TRegExpAlt(re, start, end)
+    or
+    this = TRegExpBackRef(re, start, end)
+    or
+    this = TRegExpCharacterClass(re, start, end)
+    or
+    this = TRegExpCharacterRange(re, start, end)
+    or
+    this = TRegExpNormalChar(re, start, end)
+    or
+    this = TRegExpQuote(re, start, end)
+    or
+    this = TRegExpGroup(re, start, end)
+    or
+    this = TRegExpQuantifier(re, start, end)
+    or
+    this = TRegExpSequence(re, start, end)
+    or
+    this = TRegExpSpecialChar(re, start, end)
+  }
+
+  /**
+   * Gets the outermost term of this regular expression.
+   */
+  RegExpTerm getRootTerm() {
+    this.isRootTerm() and result = this
+    or
+    result = this.getParent().(RegExpTerm).getRootTerm()
+  }
+
+  /**
+   * Holds if this term is part of a string literal
+   * that is interpreted as a regular expression.
+   */
+  predicate isUsedAsRegExp() { any() }
+
+  /**
+   * Holds if this is the root term of a regular expression.
+   */
+  predicate isRootTerm() { start = 0 and end = re.getText().length() }
+
+  override RegExpTerm getChild(int i) {
+    result = this.(RegExpAlt).getChild(i)
+    or
+    result = this.(RegExpBackRef).getChild(i)
+    or
+    result = this.(RegExpCharacterClass).getChild(i)
+    or
+    result = this.(RegExpCharacterRange).getChild(i)
+    or
+    result = this.(RegExpNormalChar).getChild(i)
+    or
+    result = this.(RegExpGroup).getChild(i)
+    or
+    result = this.(RegExpQuantifier).getChild(i)
+    or
+    result = this.(RegExpSequence).getChild(i)
+    or
+    result = this.(RegExpSpecialChar).getChild(i)
+  }
+
+  /**
+   * Gets the parent term of this regular expression term, or the
+   * regular expression literal if this is the root term.
+   */
+  RegExpParent getParent() { result.getAChild() = this }
+
+  override Regex getRegex() { result = re }
+
+  /** Gets the offset at which this term starts. */
+  int getStart() { result = start }
+
+  /** Gets the offset at which this term ends. */
+  int getEnd() { result = end }
+
+  override string toString() { result = re.getText().substring(start, end) }
+
+  /**
+   * Gets the location of the surrounding regex, as locations inside the regex do not exist.
+   * To get location information corresponding to the term inside the regex,
+   * use `hasLocationInfo`.
+   */
+  Location getLocation() { result = re.getLocation() }
+
+  /** Holds if this term is found at the specified location offsets. */
+  predicate hasLocationInfo(
+    string filepath, int startline, int startcolumn, int endline, int endcolumn
+  ) {
+    exists(int re_start, int re_end |
+      re.getLocation().hasLocationInfo(filepath, startline, re_start, endline, re_end) and
+      startcolumn = re_start + start + 1 and
+      endcolumn = re_start + end
```

Comment:
This only appears to work for the most simple cases. But covering everything is likely going to be too much work for very little benefit. Are there some common cases that it might be worth covering as well? E.g. backslashes? At the very least I think the QL code here should have a comment stating that this is a simple best-effort approach and sketch some of the things that aren't covered.
To elaborate on why this fails in general: Our offsets are necessarily in terms of the string _value_, but the location offsets are in terms of the _literal_ in the code. The most basic difference is escape sequences, a value `\` usually has the literal code representation `\\`, there can also be unicode escape sequences, and the string can have compile-time concatenation (`"a" + "b"` is a string literal with the value `ab`). String literals with compile-time concatenation can also be spread across multiple lines, which further throws off the column offsets. And then to complicate things further a string literal can also be a Java 15 text block, which has a whole set of additional rules.

Reply:
I'm aware of this issue; I'm working on handling the most common cases (i.e. backslashes).